src/main/kotlin/org/misarch/returns/service/ReturnService.kt

```diff
@@ -0,0 +1,100 @@
+        val savedReturn = repository.save(returnEntity).awaitSingle()
+        for (orderItem in orderItems) {
+            orderItem.returnedWithId = savedReturn.id
+        }
+        orderItemRepository.saveAll(orderItems).collectList().awaitSingle()
+        eventPublisher.publishEvent(
+            ReturnEvents.RETURN_CREATED,
+            ReturnDTO(savedReturn.id!!, orderId, returnInput.orderItemIds, returnInput.reason, refundedAmount)
+        )
+        return savedReturn
+    }
+
+    /**
+     * Checks that a return can be created
+     * Checks that at least one order item is provided, and all order items exist, belong to the same order,
+     * have been delivered and have not been returned yet
+     *
+     * @param returnInput the return to validate
+     * @param orderItems the order items to validate
+     * @param authorizedUser the authorized user, used for permission checking
+     */
+    private suspend fun validateReturn(
+        returnInput: CreateReturnInput, orderItems: MutableList<OrderItemEntity>, authorizedUser: AuthorizedUser
+    ) {
+        val orderId = orderItems.first().orderId
+        val order = orderRepository.findById(orderId).awaitSingle()
+        require(order.userId == authorizedUser.id) { "Only the original buyer can create returns" }
+        require(returnInput.orderItemIds.isNotEmpty()) { "At least one order item id must be provided" }
+        val missingOrderItemIds = returnInput.orderItemIds.toSet() - orderItems.map { it.id }.toSet()
+        require(missingOrderItemIds.isEmpty()) { "Order item ids $missingOrderItemIds do not exist" }
+        val returnedOrderItems = orderItems.filter { it.returnedWithId != null }
+        require(returnedOrderItems.isEmpty()) { "Order item ids ${returnedOrderItems.map { it.id }} have already been returned" }
+        require(orderItems.all { it.sentWithId != null && it.compensatableAmount != null && it.productVariantVersionId != null }) {
+            "Missing data about order items, please try again later"
+        }
+        val originalShipments = orderItems.map { it.sentWithId }.toSet()
+        val shipmentsById = shipmentRepository.findAllById(originalShipments).collectList().awaitSingle().associateBy { it.id }
+        require(shipmentsById.values.all { it.deliveredAt != null }) { "At least one order item was not delivered" }
+        require(orderItems.all { it.orderId == orderId }) { "All order items must belong to the same order" }
+        for (orderItem in orderItems) {
+            val productVariantVersion = productVariantVersionRepository.findById(orderItem.productVariantVersionId!!).awaitSingle()
+            require(productVariantVersion.canBeReturnedForDays != null) { "Order item ${orderItem.id} cannot be returned" }
+            val passedSeconds = Duration.between(shipmentsById[orderItem.sentWithId]!!.deliveredAt, OffsetDateTime.now()).toSeconds()
+            val passedDays = passedSeconds.toDouble() / Duration.ofDays(1).toSeconds()
```

Comment:
Why not use `Duration#toDays` directly?
```suggestion
            val passedDays = Duration.between(shipmentsById[orderItem.sentWithId]!!.deliveredAt, OffsetDateTime.now()).toDays()
```

Reply:
because it directly returns a long
in catalog, we allow a float, e.g. you can allow returns for 2 hours (whyever)
so I did not want to loose precision here

Reply:
Ouch. But at least it is an acceptable reason…